drivers/counter/counter_ace_v1x_art.c

```diff
@@ -0,0 +1,110 @@
+/*
+ * Copyright (c) 2022 Intel Corporation
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <device.h>
+#include <devicetree.h>
+#include <zephyr/drivers/counter.h>
+#include <soc.h>
+#include <ace_v1x-regs.h>
+#include <counter/counter_ace_v1x_art_regs.h>
+
+static struct k_spinlock lock;
+
+static void counter_ace_v1x_art_ionte_set(bool new_timestamp_enable)
+{
+	uint32_t val = sys_read32(ACE_TSCTRL);
+
+	val = ((val & ~BIT(ACE_TSCTRL_IONTE_OFFSET))
+		| (new_timestamp_enable << ACE_TSCTRL_IONTE_OFFSET));
+
+	sys_write32(val, ACE_TSCTRL);
+}
+
+static void counter_ace_v1x_art_cdmas_set(uint32_t cdmas)
+{
+	uint32_t val = sys_read32(ACE_TSCTRL);
+
+	val = ((val & ~BIT(ACE_TSCTRL_CDMAS_OFFSET))
+		| (cdmas << ACE_TSCTRL_CDMAS_OFFSET));
+
+	sys_write32(val, ACE_TSCTRL);
+}
+
+static void counter_ace_v1x_art_ntk_set(bool new_timestamp_taken)
+{
+	uint32_t val = sys_read32(ACE_TSCTRL);
+
+	val = ((val & ~BIT(ACE_TSCTRL_NTK_OFFSET))
+		| (new_timestamp_taken << ACE_TSCTRL_NTK_OFFSET));
+
+	sys_write32(val, ACE_TSCTRL);
+}
+
+static uint32_t counter_ace_v1x_art_ntk_get(void)
+{
+	return (sys_read32(ACE_TSCTRL) & (1 << ACE_TSCTRL_NTK_OFFSET)) >> ACE_TSCTRL_NTK_OFFSET;
+}
+
+static void counter_ace_v1x_art_hhtse_set(bool enable)
+{
+	uint32_t val = sys_read32(ACE_TSCTRL);
+
+	val = ((val & ~BIT(ACE_TSCTRL_HHTSE_OFFSET))
+		| (enable << ACE_TSCTRL_HHTSE_OFFSET));
+
+	sys_write32(val, ACE_TSCTRL);
+}
+
+static uint64_t counter_ace_v1x_art_counter_get(void)
+{
+	uint32_t hi0, lo, hi1;
+
+	do {
+		hi0 = sys_read32(ACE_ARTCS_HI);
+		lo = sys_read32(ACE_ARTCS_LO);
+		hi1 = sys_read32(ACE_ARTCS_HI);
+	} while (hi0 != hi1);
+
+	return (((uint64_t)hi1) << 32) | lo;
+}
+
+static int counter_ace_v1x_art_get_value(const struct device *dev, uint64_t *value)
+{
+	ARG_UNUSED(dev);
+
+	k_spinlock_key_t key = k_spin_lock(&lock);
+
+	counter_ace_v1x_art_ionte_set(1);
+	counter_ace_v1x_art_cdmas_set(1);
+
+	if (counter_ace_v1x_art_ntk_get()) {
+		counter_ace_v1x_art_ntk_set(1);
+		while (counter_ace_v1x_art_ntk_get()) {
+			k_busy_wait(10);
+		}
+	}
+
+	counter_ace_v1x_art_hhtse_set(1);
+
+	while (!counter_ace_v1x_art_ntk_get()) {
+		k_busy_wait(10);
+	}
+
+	*value = counter_ace_v1x_art_counter_get();
+
+	counter_ace_v1x_art_ntk_set(1);
+	k_spin_unlock(&lock, key);
+
+	return 0;
+}
+
+static const struct counter_driver_api ace_v1x_art_counter_apis = {
+	.get_value_64 = counter_ace_v1x_art_get_value
+};
+
+DEVICE_DT_DEFINE(DT_NODELABEL(ace_art_counter), NULL, NULL, NULL, NULL,
```

Comment:
why not define `DT_DRV_COMPAT` and use instance based definition?

Reply:
I know, that the documentation advises this kind of solution, but there is no point in doing it using instance based definition, since there will never be more than one instance of ART counter and one instance of RTC counter.